File: mac/Sources/SeanboyCore/SyncState.swift
import Foundation

/// What this device knows about the bucket from its last successful sync:
/// for each note, the remote key it lived at, the ETag it had, and the
/// note's `modifiedAt` at that moment. Comparing against this file is what
/// turns sync into a three-way merge — local change and remote change are
/// detected independently, so a skewed clock can no longer eat an edit.
public struct SyncState: Codable, Equatable, Sendable {
    public struct Entry: Codable, Equatable, Sendable {
        public var key: String
        public var etag: String
        public var modifiedAt: Date

        public init(key: String, etag: String, modifiedAt: Date) {
            self.key = key
            self.etag = etag
            self.modifiedAt = modifiedAt
        }
    }

    /// Keyed by `Note.id.uuidString` (JSON-friendly).
    public var entries: [String: Entry]

    public init(entries: [String: Entry] = [:]) {
        self.entries = entries
    }

    public subscript(id: UUID) -> Entry? {
        get { entries[id.uuidString] }
        set { entries[id.uuidString] = newValue }
    }

    // MARK: - Persistence

    public static func load(from url: URL) -> SyncState {
        guard let data = try? Data(contentsOf: url) else { return SyncState() }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return (try? decoder.decode(SyncState.self, from: data)) ?? SyncState()
    }

    public func save(to url: URL) throws {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        try encoder.encode(self).write(to: url, options: [.atomic])
    }
}

File: android/core/src/main/kotlin/com/torchcodelab/seanboy/core/SyncState.kt
package com.torchcodelab.seanboy.core

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.File
import java.time.Instant
import java.time.format.DateTimeFormatter
import java.util.UUID

/**
 * What this device knows about the bucket from its last successful sync: for
 * each note, the remote key it lived at, the ETag it had, and the note's
 * `modifiedAt` at that moment. Comparing against this is what turns sync into a
 * three-way merge — local change and remote change are detected independently,
 * so a skewed clock can no longer eat an edit.
 *
 * Kotlin port of `mac/Sources/SeanboyCore/SyncState.swift`. The in-memory shape
 * lands here; JSON persistence to app-private storage arrives with the sync
 * integration milestone.
 */
data class SyncState(
    /** Keyed by uppercase `Note.id` (JSON-friendly, matches the Mac). */
    val entries: MutableMap<String, Entry> = mutableMapOf(),
) {
    data class Entry(val key: String, val etag: String, val modifiedAt: Instant)

    operator fun get(id: UUID): Entry? = entries[NoteDocument.uuidString(id)]

    operator fun set(id: UUID, entry: Entry?) {
        val k = NoteDocument.uuidString(id)
        if (entry == null) entries.remove(k) else entries[k] = entry
    }

    // MARK: - Persistence (device-local; never synced)

    fun save(file: File) {
        val dto = Dto(entries.mapValues { EntryDto(it.value.key, it.value.etag, ISO.format(it.value.modifiedAt)) })
        file.parentFile?.mkdirs()
        val tmp = File(file.parentFile, file.name + ".tmp")
        tmp.writeText(json.encodeToString(Dto.serializer(), dto))
        if (!tmp.renameTo(file)) { file.writeText(tmp.readText()); tmp.delete() }
    }

    @Serializable
    private data class Dto(val entries: Map<String, EntryDto> = emptyMap())

    @Serializable
    private data class EntryDto(val key: String, val etag: String, val modifiedAt: String)

    companion object {
        private val ISO: DateTimeFormatter = DateTimeFormatter.ISO_INSTANT
        private val json = Json { prettyPrint = true; ignoreUnknownKeys = true }

        fun load(file: File): SyncState {
            val text = runCatching { file.readText() }.getOrNull() ?: return SyncState()
            val dto = runCatching { json.decodeFromString(Dto.serializer(), text) }.getOrNull()
                ?: return SyncState()
            val entries = dto.entries.mapValues { (_, e) ->
                Entry(e.key, e.etag, NoteDocument.parseDate(e.modifiedAt) ?: Instant.EPOCH)
            }.toMutableMap()
            return SyncState(entries)
        }
    }
}
